### PR TITLE
Support static token authentication

### PR DESCRIPTION
**Pull request by KacperKluka**
I added convenience auth methods that allow to easily use a static token authentication. Both token request and JWT are supported. The new static auth options are added to the Java API as well and are tested in integration tests.

### CODE REVIEW

File: common/src/main/java/com/ably/tracking/common/AblyHelpers.kt [hunk 1]
```diff
@@ -88,28 +88,40 @@ val Authentication.clientOptions: ClientOptions
...
         clientId = this@clientOptions.clientId
 
-        this@clientOptions.tokenRequestCallback?.let { tokenRequestCallback ->
-            authCallback = Auth.TokenCallback {
-                runBlocking {
-                    try {
-                        tokenRequestCallback(it.toTracking()).toAuth()
-                    } catch (exception: TokenAuthException) {
-                        throw exception.toAblyException()
+        this@clientOptions.tokenRequestConfiguration?.let { tokenRequestConfiguration ->
```
Comment (davyskiba) [inline common/src/main/java/com/ably/tracking/common/AblyHelpers.kt:91]: Maybe splitting those into two entirely separate cases would improve readability?
Currently it's easy to get lost with all the brackets
Screenshot 2022-11-30 at 10 01 46

File: common/src/main/java/com/ably/tracking/common/AblyHelpers.kt [hunk 2]
```diff
@@ -88,28 +88,40 @@ val Authentication.clientOptions: ClientOptions
...
         }
 
-        this@clientOptions.jwtCallback?.let { jwtCallback ->
-            authCallback = Auth.TokenCallback {
-                runBlocking {
-                    try {
-                        jwtCallback(it.toTracking())
-                    } catch (exception: TokenAuthException) {
-                        throw exception.toAblyException()
+        this@clientOptions.jwtConfiguration?.let { jwtConfiguration ->
```
Comment (davyskiba) [inline common/src/main/java/com/ably/tracking/common/AblyHelpers.kt:109]: Same as above, consider splitting lets into separate cases

**Review by davyskiba — APPROVED** [review]

File: common/src/main/java/com/ably/tracking/common/AblyHelpers.kt (hunk 2 above)
Comment (KacperKluka) [reply common/src/main/java/com/ably/tracking/common/AblyHelpers.kt]: Good idea https://github.com/ably/ably-asset-tracking-android/pull/824/commits/aa8808481405f5241bb5528a59b4de1246e3e6ff

**Review by davyskiba — APPROVED** [review]

File: core-sdk/src/main/java/com/ably/tracking/connection/ConnectionConfigurationModels.kt [hunk 3]
```diff
@@ -108,18 +118,28 @@ sealed class Authentication(
          */
         @JvmSynthetic
         fun jwt(callback: JwtCallback): Authentication =
-            JwtAuthentication(null, callback)
+            JwtAuthentication(null, JwtConfiguration(callback, null))
+
+        /**
+         * Authentication method that uses the JWT.
```
Comment (QuintinWillison) [inline core-sdk/src/main/java/com/ably/tracking/connection/ConnectionConfigurationModels.kt:124]:
```suggestion
         * Authentication method that uses a static JWT (JSON Web Token).
```

File: core-sdk/src/main/java/com/ably/tracking/connection/ConnectionConfigurationModels.kt [hunk 4]
```diff
@@ -83,7 +83,17 @@ sealed class Authentication(
          */
         @JvmSynthetic
         fun tokenRequest(callback: TokenRequestCallback): Authentication =
-            TokenAuthentication(null, callback)
+            TokenAuthentication(null, TokenRequestConfiguration(callback, null))
+
+        /**
+         * Authentication method that uses the Token Request.
```
Comment (QuintinWillison) [inline core-sdk/src/main/java/com/ably/tracking/connection/ConnectionConfigurationModels.kt:89]:
```suggestion
         * Authentication method that uses a static Token Request.
```

File: core-sdk-java/src/main/java/com/ably/tracking/java/AuthenticationFacade.kt [hunk 5]
```diff
@@ -47,6 +47,16 @@ class AuthenticationFacade {
                 callback.onCreateNewToken(it).await()
             }
 
+        /**
+         * Authentication method that uses the Token Request.
```
Comment (QuintinWillison) [inline core-sdk-java/src/main/java/com/ably/tracking/java/AuthenticationFacade.kt:51]:
```suggestion
         * Authentication method that uses a static Token Request.
```

File: core-sdk-java/src/main/java/com/ably/tracking/java/AuthenticationFacade.kt [hunk 6]
```diff
@@ -58,5 +68,15 @@ class AuthenticationFacade {
             Authentication.jwt {
                 callback.onCreateNewToken(it).await()
             }
+
+        /**
+         * Authentication method that uses the JWT.
```
Comment (QuintinWillison) [inline core-sdk-java/src/main/java/com/ably/tracking/java/AuthenticationFacade.kt:73]:
```suggestion
         * Authentication method that uses a static JWT (JSON Web Token).
```

**Review by QuintinWillison — APPROVED** [review]
Nice, thanks. I've added a couple of minor suggestions.

**Pull request merged**